Add selector for `compile_commands.json`

Discookie opened the pull request:
Closes #15.

Currently the database setup dialog has 3 options:
* `Run CodeChecker log` creates a terminal, with a pre-defined `CodeChecker log` command already written there.
* `Locate` shows a file selector to select the database's path, eg. for CMake with a `/build` folder
* `Don't show again` never shows the window again on folder open.

The dialog shows when compile_commands.json is not found.
There's also a button in the Overview in an empty folder to select its path.

Currently there's no separation between missing the metadata.json and missing the compile_commands.json in the sidepanel.
I'd like to add a separate sidebar panel option for it before merging.

Discookie commented:
Added the sidepanel logic, should be ready now.

csordasmarton commented on package.json:
```diff
@@ -39,6 +39,10 @@
         "title": "CodeChecker: Previous reproduction step"
       },
+      {
+        "command": "codechecker.editor.showSetupDialog",
```
Can you please update the main readme file with these new commands and options?

csordasmarton commented on package.json:
```diff
@@ -72,6 +76,19 @@
+            "string",
+            "null"
+          ],
+          "description": "Path to a custom `compile_commands.json`, in case of a custom build system",
```
```suggestion
          "description": "Path to a custom compilation database, in case of a custom build system",
```

Discookie replied:
Fixed in all applicable places.

csordasmarton commented on package.json:
```diff
@@ -72,6 +76,19 @@
+        },
+        "codechecker.editor.showDatabaseDialog": {
+          "type": "boolean",
+          "description": "Show a dialog if `compile_command.json` is not found",
```
```suggestion
          "description": "Show a dialog if compilation database is not found",
```

csordasmarton commented on src/backend/executor/process.ts:
```diff
@@ -130,6 +153,41 @@ export class ExecutorProcess {
+            ?? 'CodeChecker';
+        const ccFolder = getConfigAndReplaceVariables('codechecker.backend', 'outputFolder')
+            ?? path.join(workspaceFolder, '.codechecker');
+        const ccArguments = getConfigAndReplaceVariables('codechecker.executor', 'arguments') ?? '';
```
These arguments are options for the `CodeChecker analyze` command and not for the `CodeChecker log` command. For example I set this argument to `--ctu` but the CodeChecker log command doesn't have such an option and it will fail.

So please create a separate option for it.
We can name the two arguments in the following names:
- Analysis Arguments
- Log arguments

It would also be good to create a separate command to show the full CodeChecker log command similar what we did for the CodeChecker analyze command.

Discookie replied:
Fixed by simply removing all non-essential arguments.

I don't think there's a need for Log arguments, since the user is able to edit the command-line before running it.

The command is only meant as a guideline after all, because the user has to fill in their own build system (if different from make) per project anyways.

csordasmarton commented on src/backend/executor/process.ts:
```diff
@@ -130,6 +153,41 @@ export class ExecutorProcess {
+        const ccThreads = workspace.getConfiguration('codechecker.executor').get<string>('threadCount');
+
+        // Use a predefined path here
+        const ccCompileCmd = path.join(ccFolder, 'compile_cmd.json');
```
```suggestion
        const ccCompileCmd = path.join(ccFolder, 'compile_commands.json');
```

Discookie replied:
Changed, thanks!

csordasmarton commented on src/backend/executor/process.ts:
```diff
@@ -130,6 +153,41 @@ export class ExecutorProcess {
+            `--output "${ccCompileCmd}"`,
+            `${ccThreads ? '-j ' + ccThreads : ''}`,
+            `${ccArguments}`,
+            '--build "make"'
```
This has to be configurable because not all the projects runs simply make. For example if someone wants to compile a single make target it is not possible for now.

Above I mentioned to create a separate argument for analysis and in this case it could be the default value of it.

Discookie replied:
It's configurable, as the user is presented a terminal where they can edit the command before running it.
The command here is only meant as a guideline, and 'make' is the one that's most common, I think.

csordasmarton commented on src/backend/executor/process.ts:
```diff
@@ -130,6 +153,41 @@ export class ExecutorProcess {
+        return [
+            `${ccPath} log`,
+            `--output "${ccCompileCmd}"`,
+            `${ccThreads ? '-j ' + ccThreads : ''}`,
```
CodeChecker log command doesn't have such option.

```suggestion
```

Discookie replied:
Removed, along with most other arguments.

csordasmarton commented on src/sidebar/views/overview.ts:
```diff
@@ -16,91 +16,87 @@ export class OverviewItem {
+            }),
+            new OverviewItem('——'),
+            new OverviewItem('compile_commands.json not found.'),
+            new OverviewItem('Show database setup dialog...', {
```
I clicked on this button first, created the compilation database and if I click to this button again it will show the same pop-up window:
![image](https://user-images.githubusercontent.com/6695818/141765944-c51c0175-2472-4ca6-829d-e473c613981a.png)
It is a little bit missleading because I already have a compilation database.

csordasmarton replied:
We have to show this item above the `Reload CodeChecker metadata` item because it is more important and it would be better to rename this item something to similar:
```suggestion
            new OverviewItem('Init compilation database', {
```

Discookie replied:
Fixed the button order, and added a new text when there's already a database selected.

I'd rather not rename the button, as I refer to the dialog as the `database setup dialog` in the settings and the README.

csordasmarton commented on src/sidebar/views/overview.ts:
```diff
@@ -16,91 +16,87 @@ export class OverviewItem {
+                command: 'codechecker.backend.reloadMetadata',
+            }),
+            new OverviewItem('——'),
+            new OverviewItem('compile_commands.json not found.'),
```
Why do we need this item? I created the compilation database but this item is still there and it is missleading.

Discookie replied:
It should automatically hide itself when the user selects a database.
If the sidepanel doesn't reload when the database is created, that's a bug.

I should add watches to each of the database paths, and fire an event when one's changed.

Discookie replied:
Added watches for the files, so the sidebar should update when `CodeChecker log` finished running.

csordasmarton commented on src/sidebar/views/overview.ts:
```diff
@@ -16,91 +16,87 @@ export class OverviewItem {
+            new OverviewItem('Loading overview, please wait...')
+        ],
+        'notFound': [
+            new OverviewItem('CodeChecker run not found.'),
+            new OverviewItem('Run CodeChecker, reload metadata,'),
+            new OverviewItem('or set the output folder to get started')
```
```suggestion
            new OverviewItem('Note: CodeChecker is not run yet. Please use commands below to get started.')
```

With the `Note` prefix we will indicate that this is just a note text and it is not clickable.

Discookie replied:
The sidebar items don't wrap, so we're limited by line width here.
And since there are no buttons in that group, even when statistics are displayed, I'm not in favour of putting `Note` everywhere.

Maybe in the future the same kind of dialog could be shown for CodeChecker analyze, or perhaps some other way of walking the user through setting it up.

csordasmarton commented on src/sidebar/views/overview.ts:
```diff
@@ -16,91 +16,87 @@ export class OverviewItem {
+            }),
+        ],
+        'ccNotFound': [
+            new OverviewItem('Reload CodeChecker metadata', {
```
I clicked on the  `Show database setup dialog...` item, initialized a compilation database but the CodeChecker view still shows me that the compilation database is not shown even if I click on the File view and go back to the CodeChecker view:
![image](https://user-images.githubusercontent.com/6695818/141768684-3a2c283d-2522-4ac2-84e2-4c2f26e2bff9.png)

If I click on the Reload CodeChecker metadata button the view will be refreshed and it will show me the following buttons:
![image](https://user-images.githubusercontent.com/6695818/141769032-42fa4e00-a111-4ec1-85b2-a9a131d7cac2.png)

Can not we show this automatically.

Discookie replied:
Added watches, so the sidebar should update automatically even when running Log.

Discookie commented:
Added watchers for all possible database paths, and the sidepanel should automatically refresh when at least one of them is found.
Fixed most of the other issues as well.

csordasmarton reviewed (approved):
LGTM!
